common/types.go

```diff
@@ -0,0 +1,182 @@
+type FileData struct {
+	Package string // Name of the package
+	Name    string // Name of the file (basename)
+	Count   int64  // Number of statements in the file
+	Exec    int64  // Number of statements in the file that were executed
+}
+
+// Coverage reports the coverage of the file as a float.
+func (fd FileData) Coverage() float64 {
+	// Avoid divide-by-zero
+	if fd.Count <= 0 {
+		return 1.0
+	}
+
+	return float64(fd.Exec) / float64(fd.Count)
+}
+
+// Handle reports a handle for the FileData record.  This is typically
+// the full file name, including package, but it could be just the
+// package name, or the empty string.
+func (fd FileData) Handle() string {
+	if fd.Package == "" && fd.Name == "" {
+		return ""
+	}
+
+	return fmt.Sprintf("%s/%s", fd.Package, fd.Name)
+}
+
+// String reports the coverage of the entity described by the FileData
+// object.
+func (fd FileData) String() string {
+	handle := fd.Handle()
+	if handle == "" {
+		return fmt.Sprintf("%d statements out of %d covered; overall coverage: %.1f%%", fd.Exec, fd.Count, fd.Coverage()*100.0)
+	}
+
+	return fmt.Sprintf("%s: %d statements out of %d covered; coverage: %.1f%%", handle, fd.Exec, fd.Count, fd.Coverage()*100.0)
+}
+
+// DataSet is a list of FileData instances.  It implements
+// sort.Interface, and thus is capable of being sorted by handle.
+type DataSet []FileData
+
+// Merge is a utility function that merges a list of FileData
+// instances with another FileData list.  It ensures that the Count
+// fields are the same and picks an Exec field.  It returns the
+// resulting list, along with a list of FileData where Count did not
+// match; the order will match the elements of a first, followed by
+// any elements of b which did not appear in a.
+func (ds DataSet) Merge(other DataSet) (DataSet, DataSet) {
+	// Set up our index and result sets
+	idx := map[string]map[string]int{}
+	var result DataSet
+	var conflict DataSet
+
+	// Copy ds into the result set
+	i := 0
+	for _, fd := range ds {
+		// Mark it seen
+		if _, ok := idx[fd.Package]; !ok {
+			idx[fd.Package] = map[string]int{}
+		}
+		idx[fd.Package][fd.Name] = i
+
+		// Save it to the result set
+		result = append(result, fd)
+		i++
+	}
+
+	// Now walk through other
+	for _, fd := range other {
+		// Have we seen it?
+		if _, ok := idx[fd.Package]; !ok {
+			idx[fd.Package] = map[string]int{}
+		}
+		if j, ok := idx[fd.Package][fd.Name]; ok {
+			// Are the counts consistent?
+			if result[j].Count != fd.Count {
+				conflict = append(conflict, fd)
+			} else if result[j].Exec == 0 {
+				result[j].Exec = fd.Exec
+			}
+			continue
+		}
+
+		// Add the element to the result set
+		idx[fd.Package][fd.Name] = i
+		result = append(result, fd)
+		i++
+	}
+
+	return result, conflict
+}
+
+// Reduce is a utility function that reduces a list of FileData
+// instances into a list of FileData instances that only contain
+// counts for packages.
+func (ds DataSet) Reduce() DataSet {
```

Comment:
It's times like this I wish Golang had proper map/reduce functions in the stdlib 😩

Reply:
Yeah, well; maybe once they actually get generics in...